### PR TITLE
MAINT: `_lib`: update `_docscrape` from upstream

### PR DESCRIPTION
**Pull request by lucascolley**
[skip cirrus]

#### Reference issue
n/a

#### What does this implement/fix?

Brings `_lib._docscrape` up to date with upstream.

#### Additional information
This is used in `stats._axis_nan_policy`.

Accidentally pushed a branch to the main repo again, will fix my codespaces setup.

### CODE REVIEW

File: tools/lint.toml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 exclude = [
     "scipy/_build_utils/tempita/*",
+    "scipy/_lib/_docscrape",
```
Comment (lucascolley) [inline tools/lint.toml:3]:
```suggestion
    "scipy/_lib/_docscrape.py",
```

**Comment by lucascolley** [comment]
@mdhaber the first commit passed all CI apart from the linter and the second commit makes the linter pass. I don't know whether you want to check the behaviour with `stats._axis_nan_policy` or whether we are happy that this is safe to do.

**Comment by mdhaber** [comment]
I would be able to tell by looking at the docs, but there is no docs artifact.

**Comment by lucascolley** [comment]
> I would be able to tell by looking at the docs, but there is no docs artifact.

https://output.circle-artifacts.com/output/job/65899f3e-888a-4b96-b87c-939f48a98d65/artifacts/0/html/index.html?pr=21215

> I'd also like to compare old _docsrape + black formatting with this version, since it seems like that would account for most changes. Is it convenient for you to generate that diff (if you ever use Black already)?

Okay, I can look at generating this diff, if that is preferable to you taking a look through the ~10 commits since you added the script at https://github.com/numpy/numpydoc/commits/main/numpydoc/docscrape.py.

**Review by mdhaber — COMMENTED** [review]
I compared the original version of this file against our current version to find things we fixed locally. Mostly we did f-string conversions and other simple things that were also done upstream, but there were two issues that we might want to fix upstream (described inline). 

I also compared our current version formatted with black against the PR version, and the upstream fixes don't look unusual, so I assume they're useful.

I also looked at how the documentation renders, and it looks OK, so this is still doing its job.

Approved once we consider these two inline comments.

**Comment by lucascolley** [comment]
Thanks Matt, PR opened upstream. The linter did catch the missing `stacklevel` in the first commit, but I disabled linting for this file in the second commit - my understanding is that we leave checking the quality of vendored code to the upstream repos.

**Comment by mdhaber** [comment]
Ah, got it. Do you want to make any changes here since they will probably be fixed, or merge now?

**Comment by lucascolley** [comment]
no-rush on merge here, I'll make the changes here at some point (maybe before or after the upstream PR is in)

**Comment by mdhaber** [comment]
CI failure is also in gh-21167, so unrelated Looks like the changes made it in upstream, too, so I'll go ahead and merge.

**Pull request merged**